### PR TITLE
Working Workflow Example

### PR DESCRIPTION
**Pull request by rosecers**
Some notes:
- because this is living in the kernel-tutorials repo, I expanded the sections on methods covered _in_ kernel tutorials, but left the methods not covered in kernel-tutorials somewhat cursory. In other words, cells for librascal and chemiscope are somewhat succinct, although this can be brought up for debate.
- I've left the notebook unstripped so that you don't have to run it to see the output. We can strip upon merge.
- There is the resulting chemiscope file for viewing under `notebooks/kpcovr-chemiscope.json`

### CODE REVIEW

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 1]
```diff
@@ -0,0 +1,1002 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#!/usr/bin/env python3\n",
+    "import sys\n",
+    "import numpy as np\n",
+    "from sklearn.linear_model import RidgeCV\n",
+    "\n",
+    "\n",
+    "def calc_rmse(y1, y2, y_scaler, rounding=4):\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb:42]: This is somehow self-explanatory, but a short comment/docstring for parameters would be nice

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 2]
```diff
@@ -0,0 +1,1002 @@
...
+    "\n",
+    "# Librascal\n",
+    "from rascal.representations import SphericalInvariants as SOAP"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Because we are demonstrating with an environment-centred property, we pull the properties from the frame.arrays object.**"
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]:
```suggestion
    "**Because we are using environment-centred properties in this demonstration, we pull the properties from the frame.arrays object.**"
```

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 3]
```diff
@@ -0,0 +1,1002 @@
...
+   "source": [
+    "N = 20\n",
+    "input_file = \"../datasets/CSD-1000R.xyz\"\n",
+    "properties = [\"CS_local\", \"CS_total\"]\n",
+    "target = \"atom\"\n",
+    "\n",
+    "# Read the first N frames of CSD-500\n",
+    "frames = np.asarray(io.read(input_file, index=\":{}\".format(N)), dtype=object)\n",
+    "\n",
+    "# Extract chemical shifts\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]: This comment do not match the code

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 4]
```diff
@@ -0,0 +1,1002 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "There are 2520 features\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Compute SOAPs (from librascal tutorial)\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]: Maybe add a link to said tutorial?

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 5]
```diff
@@ -0,0 +1,1002 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# # Atomistic structure manipulation\n",
+    "# from ase import io\n",
+    "\n",
+    "# # Import SOAP vectors from numpy file\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]:
```suggestion
    "# # read/load SOAP vectors from numpy file\n",
```

`import` has a separate meaning in Python, so I think something like read or load would remove a possible source of confusion.

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 6]
```diff
@@ -0,0 +1,1002 @@
...
+    "# Y_raw = data['Y'] # property vectors (n_samples x n_properties)\n",
+    "# target = data['target']\n",
+    "# properties = data['property_names'] # names of your properties\n",
+    "\n",
+    "# hyper_file = 'soap_hypers.npz'\n",
+    "# hypers = np.load(hyper_file)\n",
+    "\n",
+    "# num_features = X_raw.shape[1]\n",
+    "\n",
+    "# # If you end up wanting a chemiscope, you'll need to import your frames as well\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]:
```suggestion
    "# # If you end up wanting a chemiscope, you'll need to read/load your frames as well\n",
```

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 7]
```diff
@@ -0,0 +1,1002 @@
...
+    "## Scale and Center data"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from skcosmo.preprocessing import StandardFlexibleScaler as SFS\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]: I don't like renaming imports, it makes it harder for a reader of the code to find the corresponding documentation, since they have to find the actual name first.

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 8]
```diff
@@ -0,0 +1,1002 @@
...
+    "\n",
+    "First, we will take a materials dataset, in the form of cartesian coordinates, and translate it into SOAP descriptors using `librascal`, as was covered briefly in [**X_ImportingData.ipynb**](X_ImportingData.ipynb).\n",
+    "\n",
+    "Next, we'll reduce the feature space of these SOAP vectors using PCov-CUR, which is covered in [**5_CUR.ipynb**](5_CUR.ipynb). \n",
+    "\n",
+    "From here, we'll select an _active_ set of samples using PCov-CUR and use them to build a set of sparse kernels, as was introduced [**4_SparseKernelMethods.ipynb**](4_SparseKernelMethodsipynb), and use this sparse kernel to build a `SparseKPCovR` map.\n",
+    "\n",
+    "Finally, we will port this `SparseKPCovR` into a `chemiscope` visualization (covered in [**1_LinearMethods.ipynb**](1_LinearMethods.ipynb)), which can be loaded on [chemiscope.org](chemiscope.org).\n",
+    "\n",
+    "With each of the sections, we'll provide a \"fast-forward\" cell, where you can skip the section by un-commenting and executing the cell. For example, if you have pre-computed your SOAP vectors and would like to simply import them, skip section 1 and execute the cell under `1.1 Skip Computing SOAPS`."
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]: Maybe add a section on "dependencies you'll need", giving a `pip install` command? Something like `pip install sklearn numpy matplotlib scipy skcosmo ase chemiscope git+https://github.com/cosmo-epfl/librascal#e78f1c69`

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 9]
```diff
@@ -0,0 +1,1002 @@
...
+     ]
+    }
+   ],
+   "source": [
+    "r = RidgeCV(cv=4, fit_intercept=False, alphas=np.logspace(-8, 2, 10))\n",
+    "r.fit(X_raw_train, Y_train)\n",
+    "\n",
+    "Yp_train = r.predict(X_raw_train)\n",
+    "Yp_test = r.predict(X_raw_test)\n",
+    "rmse = round(\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]: why no use `calc_rmse` here?

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 10]
```diff
@@ -0,0 +1,1002 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Determine the active kernel\n",
+    "**Keep in mind that this example is originally written with an environment-centred property, and therefore we can use an environment kernel.**\n",
+    "\n",
+    "Until you can specify an active set with sklearn, we'll construct a sparse nystroem approximation in a similar manner, and replace this with the sklearn functionality when it becomes available."
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]: This sentence is very specific to the time this notebook is written, should we remove it? 

We are not trying to have a notebook that is using sklearn as much as possible.

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 11]
```diff
@@ -0,0 +1,1002 @@
...
+   "outputs": [],
+   "source": [
+    "active_kernel = pairwise_kernels(\n",
+    "    X_active, metric=kernel, filter_params=True, **kernel_params\n",
+    ")\n",
+    "\n",
+    "# By including the right singular vectors into the RKHS features, we normalize and center the RKHS.\n",
+    "U, S, V = svd(active_kernel)\n",
+    "S = np.maximum(S, 1e-12)\n",
+    "normalization_ = np.dot(U / np.sqrt(S), V)\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]:
```suggestion
    "normalization = np.dot(U / np.sqrt(S), V)\n",
```

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 12]
```diff
@@ -0,0 +1,1002 @@
...
+    "components_ = X_active.copy()\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]:
```suggestion
    "components = X_active.copy()\n",
```

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 13]
```diff
@@ -0,0 +1,1002 @@
...
+    "\n",
+    "Phi_train = np.dot(\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb:720]: This part of the code is not very clear. Amongst other things, Phi is not defined, the `1e-12` in `S = np.maximum(S, 1e-12)`  is a bit arbitrary, and the use of SVD to build a normalization matrix is not explained. Is this talked about in the previous notebooks?

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 14]
```diff
@@ -0,0 +1,1002 @@
...
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "kpcovr_hypers = dict(mixing=0.5, n_components=5)\n",
+    "\n",
+    "kpcovr = KPCovR(kernel=\"linear\", center=True, **kpcovr_hypers)\n",
+    "kpcovr.fit(\n",
+    "    Phi_train, Y_train, \n",
+    "    Yhat=Yp_train # This will throw an error if Y has not been estimated.\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]: not sure what the comment means. How can I know/make sure Y has been estimated, and what does "estimate Y" means in this context? 

If this is all material covered in previous notebooks, then my question becomes how much do we want to couple this notebook to previous ones?

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 15]
```diff
@@ -0,0 +1,1002 @@
...
+    "# print(\n",
+    "#     f\"By comparison, a regression on {kpcovr.n_components} KPCA components gives an RMSE of {rmse}ppm.\"\n",
+    "# )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Exporting to a Chemiscope"
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]:
```suggestion
    "# Exporting to chemiscope"
```

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 16]
```diff
@@ -0,0 +1,1002 @@
...
+    "import chemiscope\n",
+    "\n",
+    "output_name = \"kpcovr-chemiscope.json\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll write one output combining the training set and the testing set."
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]:
```suggestion
    "Chemiscope is an online interactive visualizer that allows you to look at the projection in relation with the structures. Here, we'll create a chemiscope input file combining the training set and the testing set."
```

File: notebooks/6_Putting_it_All_Together.ipynb [hunk 17]
```diff
@@ -0,0 +1,1002 @@
...
+    "is_active[i_train[sample_idx]] = 1\n",
+    "\n",
+    "n_cur = 10\n",
+    "cur_features = np.zeros((X_raw.shape[0], n_cur))\n",
+    "cur_features[i_test] = X_raw_test[:, cur_idx[:n_cur]]\n",
+    "cur_features[i_train] = X_raw_train[:, cur_idx[:n_cur]]\n",
+    "\n",
+    "chemiscope_properties = {\n",
+    "    **{\n",
+    "        f\"kpcovr_{i}\": {\"target\": target, \"values\": T[:, i]}\n",
```
Comment (Luthaf) [inline notebooks/6_Putting_it_All_Together.ipynb]: this is done automatically in chemiscope.write_input if you pass

```py
chemiscope_properties = {
    "kpcovr": {
        "target": "atom", 
        "values": T[:, :kpcovr.n_components]
    },
    "cur_feature": {
        "target": "atom", 
        "values": cur_features[:, :n_cur]
   },
    **{
        prop_name: {"target": target, "values": Y[:, i]}
        for i, prop_name in enumerate(properties)
    },
}
```

**Review by Luthaf — COMMENTED** [review]
I have mostly small comments, as well as a higher level one. 

This notebook is very good as a "you have done all of kernel-tutorials, now here is a end-to-end example of how you could use it" notebook, which is a different use case from the "here is how you get from a structure to a chemiscope input" which I had in mind for https://github.com/cosmo-epfl/chemiscope/issues/114. 

In particular, the use of (PCov-)FPS, (PCov-)CUR, and sparse methods is adding a lot of complexity for people that might just want to try out chemiscope quickly. The notebook also assumes knowledge of SOAP and descriptors, and do not introduce the reason behind dimensionality reduction.

I think the question here is how much do we want to couple this to the rest of kernels-tutorials/skcosmo. Should someone be able to read/use this notebook without reading the rest of kernel-tutorials?

**Comment by rosecers** [comment]
@Luthaf As for your larger comment, I think that's the conversation you, me, @max-veit, and @ceriottm need to have. If we want a front-to-back ase-to-chemiscope pipeline, where should it live? Is it meant to be internal to the group or publicly available? 
As I said, I went heavy on the scikit-cosmo-like functionalities because we're in kernel-tutorials. A possible solution is to have three notebooks all doing the same thing, but with different emphases (librascal, scikit-cosmo/KT, chemiscope), living in the three packages. This is a bit dangerous though, due to the constantly updating nature of the packages.

I made this notebook 1) to start this conversation, 2) because I think regardless of the answer, it's a nice "bow" to K-T, and 3) because until we can figure out a more centralized example, it can serve as a "until-then" solution for group members wanting a front-to-back solution. What do you think?

**Comment by Luthaf** [comment]
Fully agree on all points, and we can move forward with this notebook as a "bow" for kernels-tutorials regardless of the outcome of the other discussion!

Personally, I think having a separate "this is the simplest way to create a chemiscope input" example would be very valuable; and could point to the notebook introduced in this PR as a good point of "here is how you can improve the projection a lot".

File: notebooks/6_Putting_it_All_Together.ipynb (hunk 1 above)
Comment (rosecers) [reply notebooks/6_Putting_it_All_Together.ipynb:42]: Addressed in most recent commit!

File: notebooks/6_Putting_it_All_Together.ipynb (hunk 2 above)
Comment (rosecers) [reply notebooks/6_Putting_it_All_Together.ipynb]: Addressed in most recent commit!

File: notebooks/6_Putting_it_All_Together.ipynb (hunk 4 above)
Comment (rosecers) [reply notebooks/6_Putting_it_All_Together.ipynb]: Addressed in most recent commit!

File: notebooks/6_Putting_it_All_Together.ipynb (hunk 3 above)
Comment (rosecers) [reply notebooks/6_Putting_it_All_Together.ipynb]: Addressed in most recent commit!

File: notebooks/6_Putting_it_All_Together.ipynb (hunk 5 above)
Comment (rosecers) [reply notebooks/6_Putting_it_All_Together.ipynb]: Addressed in most recent commit!

File: notebooks/6_Putting_it_All_Together.ipynb (hunk 14 above)
Comment (rosecers) [reply notebooks/6_Putting_it_All_Together.ipynb]: Addressed in most recent commit!

File: notebooks/6_Putting_it_All_Together.ipynb (hunk 15 above)
Comment (rosecers) [reply notebooks/6_Putting_it_All_Together.ipynb]: Addressed in most recent commit!

File: notebooks/6_Putting_it_All_Together.ipynb (hunk 16 above)
Comment (rosecers) [reply notebooks/6_Putting_it_All_Together.ipynb]: Addressed in most recent commit!

File: notebooks/6_Putting_it_All_Together.ipynb (hunk 17 above)
Comment (rosecers) [reply notebooks/6_Putting_it_All_Together.ipynb]: Addressed in most recent commit!
